Modifying how data is written in to test app

barryskal opened the pull request:
I was having some issues running the integration tests, in particular with TestHaproxy where data is written in to the testapp. What I was seeing was that the second time data was written in to the test app the write would not complete before the test app was deployed. 

I've changed the method for writing this data file to utilise a different method of writing to this file to make sure that it's been completed before deployment by utilising the fact that os.Rename is atomic.

jsuresh commented on integration_test.go:
```diff
@@ -375,10 +375,24 @@ func TestStop(t *testing.T) {
 	if err != nil {
-		t.Fatalf("write file: %s\n", err)
+		t.Fatalf("could not create temp file: %s\n", err)
 	}
```
defer os.Remove(tempFile.Name())

By the doco, if the file doesn't exist, this will do nothing

jsuresh commented on integration_test.go:
```diff
@@ -375,10 +375,24 @@ func TestStop(t *testing.T) {
+
+	_, err = tempFile.Write([]byte(data))
+	if err != nil {
+		os.Remove(tempFile.Name())
```
delete

jsuresh commented on integration_test.go:
```diff
@@ -375,10 +375,24 @@ func TestStop(t *testing.T) {
+	tempFile.Close()
+	err = os.Rename(tempFile.Name(), "testapp/data/file")
+	if err != nil {
+		os.Remove(tempFile.Name())
```
delete

jsuresh commented:
Looks pretty good, a few comments on go style as to how to use defer etc. Chat to ben if you are unsure (he just did exactly this feature)

Otherwise LGTM % comments

jsuresh commented on integration_test.go:
```diff
@@ -375,10 +375,24 @@ func TestStop(t *testing.T) {
 }
 
 func writeDataIntoTestapp(t *testing.T, data string) {
```
A better factoring would be to name this function `atomicWriteFile(t, data, filename)` Then it's not too tightly coupled with just writing a single file into the testapp.

Up to you if you want to make this change before commiting

barryskal replied:
Makes sense to me. I'll make the change

jsuresh commented:
LGTM

latos commented on integration_test.go:
```diff
@@ -374,11 +374,33 @@ func TestStop(t *testing.T) {
+
+	_, err = tempFile.Write([]byte(data))
+	if err != nil {
+		fmt.Errorf("Could not write to temp file: %s\n", err)
```
I think you've missed a return statement here

barryskal replied:
yep, sorry about that

latos commented on integration_test.go:
```diff
@@ -374,11 +374,33 @@ func TestStop(t *testing.T) {
+	err = os.Rename(tempFile.Name(), filename)
 	if err != nil {
-		t.Fatalf("write file: %s\n", err)
+		fmt.Errorf("Could not move temp file to testapp: %s\n", err)
```
And here

barryskal replied:
and again

latos commented on integration_test.go:
```diff
@@ -374,11 +374,33 @@ func TestStop(t *testing.T) {
+}
+
+func writeDataIntoTestapp(t *testing.T, data string) {
+	err := atomicWriteFile(t, data, "testapp/data/file")
```
fix indent

test the error conditions maybe?

jsuresh replied:
@barryskal you should run everything through go fmt (I've got a vim plugin that does this on save automatically)

barryskal replied:
Ah, that's much easier. Thank you 👍 
@latos I'm not sure exactly what you mean by testing the error conditions manually, we can chat about it when I see you later this week

barryskal commented:
Fixed up the issues from the previous comments. I noticed that the atomicWriteFile had a parameter of type *testing.T which wasn't used, so I removed it. 
Tested the method manually using the following code 
```
func main() {
  err := atomicWriteFile("hello", "/tmp/test/okFile")
  if err != nil {
    fmt.Printf("Failed test 1: %s\n", err)
  } else {
    fmt.Printf("Passed test 1\n")
  }


  err = atomicWriteFile("hello", "/tmp/test/no_folder_here/okFile")
  if err == nil {
    fmt.Printf("Failed test 2: No error\n")
  } else {
    if strings.Contains(err.Error(), "Could not move temp file") {
      fmt.Printf("Passed test 2\n")
    } else {
      fmt.Printf("Failed test 2: expected error containing 'Could not move temp file': got %s\n", err)
    }
  }

  tempDir := os.TempDir()
  os.Chmod(tempDir, 0500)
  defer os.Chmod(tempDir, 0700)
  err = atomicWriteFile("hello", "/tmp/test/okFile")
  if err == nil {
    fmt.Printf("Failed test 3: No error\n")
  } else {
    if strings.Contains(err.Error(), "could not create temp file") {
      fmt.Printf("Passed test 3, error is: %s\n", err)
    } else {
      fmt.Printf("Failed test 3: expected error containing 'could not create temp file': got %s\n", err)
    }
  }

}
```

Output was:
```
Passed test 1
Passed test 2
Passed test 3, error is: could not create temp file: open /var/folders/8j/b7rql61n3fs3l4zxvt016d3m0000gn/T/tempData538439475: permission denied
```

jsuresh commented on integration_test.go:
```diff
@@ -374,10 +374,32 @@ func TestStop(t *testing.T) {
+
 func writeDataIntoTestapp(t *testing.T, data string) {
-	err := ioutil.WriteFile("testapp/data/file", []byte(data), os.FileMode(0644))
+	err := atomicWriteFile(data, "testapp/data/file")
```
go pattern here is

```
if err := atomicWriteFile(..); err != nil {
}
```

This keeps the error variable scoped to the 'if' statement

barryskal replied:
Done

jsuresh commented:
Final style comment, once fixed LGTM, push and close of this PR
